
library(Seurat)
library(dplyr)
library(patchwork)
library(ggplot2)
library(sf)
library(glue)
library(celldex)
library(SingleR)
library("singleCellTK")
out_path='Results'
file_path="/Users/JJOHN41/Desktop/projects/test_data/single_cell_rnaseq/data/ctrl_raw_feature_bc_matrix/"
project_name<-"pbmc3k"


# Load dataset
seurat_data <- Read10X(file_path)

# Initialize the Seurat object with the raw (non-normalized data).
seurat_obj <- CreateSeuratObject(counts = seurat_data,
                                project = project_name,
                                min.cells = 3,
                                min.features =50,assay = "RNA")


head(seurat_obj@meta.data)



#add the doublet annotation generated by scrublet
doublets <- read.table(glue('{file_path}scrublet_calls.tsv'),header = T,row.names = 1)
seurat_obj <- AddMetaData(seurat_obj,doublets)
head(seurat_obj[[]])


# MT RNA percentage
seurat_obj[["percent.mt"]] <- PercentageFeatureSet(seurat_obj, pattern = "^MT-")
# Ribosomal RNA percentnage
seurat_obj[["percent.ribo"]] <- PercentageFeatureSet(seurat_obj, pattern = "^RP[SL]")
#iggenes
seurat_obj[["percent.igg"]] <- PercentageFeatureSet(seurat_obj, pattern = "^HLA-|IG[HJKL]")

##


create_plots <- function(seurat_obj, out_path, prefix, feature_pairs) {
        # Create directory if it doesn't exist # nolint
        if (!dir.exists(glue("{out_path}/Plots"))) 
                            dir.create(glue("{out_path}/Plots"), recursive = TRUE) # nolint

        # Save violin plot as PNG # nolint
        png(file = file.path(glue("{out_path}/Plots/{prefix}_Violinplots.png")), units = "in", width = 12, height = 12, res = 800) 
        print(VlnPlot(seurat_obj, 
                features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.ribo"),
                ncol = 4, pt.size = 0.1) + theme(plot.title = element_text(size = 10))) 
        dev.off()

        for (pair in feature_pairs) {
                filename <- glue("{out_path}/Plots/{prefix}_{pair[1]}_vs_{pair[2]}_scatter.png") 
                # Create scatter plot and save as PNG 
                png(file = filename, units = "in", width = 12, height = 12, res = 800) 
                print(FeatureScatter(seurat_obj, feature1 = pair[1], feature2 = pair[2])) 
                dev.off()
         }
        } 



# Usage example:
feature_pairs <- list(
  c("nCount_RNA", "percent.mt"),
  c("nCount_RNA", "nFeature_RNA"),
  c("nCount_RNA", "percent.ribo"),
  c("percent.ribo", "percent.mt"),
  c("nFeature_RNA","percent.mt")
)

create_plots(seurat_obj, "Results", "Rawdata", feature_pairs)


minimum_nfeature=200
maximum_nfeature=2000
percent_mt=5
seurat_obj_filt <- subset(seurat_obj, subset = nFeature_RNA > minimum_nfeature & nFeature_RNA < maximum_nfeature & percent.mt < percent_mt)
create_plots(seurat_obj_filt, "Results", "Filtered", feature_pairs)


##Duplite base filtering
seurat_obj[['QC']] <- ifelse(seurat_obj@meta.data$Is_doublet == 'True','Doublet','Pass')
seurat_obj[['QC']] <- ifelse(seurat_obj@meta.data$nFeature_RNA < 500 & seurat_obj@meta.data$QC == 'Pass','Low_nFeature',seurat_obj@meta.data$QC)
seurat_obj[['QC']] <- ifelse(seurat_obj@meta.data$nFeature_RNA < 500 & seurat_obj@meta.data$QC != 'Pass' & seurat_obj@meta.data$QC != 'Low_nFeature',paste('Low_nFeature',seurat_obj@meta.data$QC,sep = ','),seurat_obj@meta.data$QC)
seurat_obj[['QC']] <- ifelse(seurat_obj@meta.data$percent.mt > 15 & seurat_obj@meta.data$QC == 'Pass','High_MT',seurat_obj@meta.data$QC)
seurat_obj[['QC']] <- ifelse(seurat_obj@meta.data$nFeature_RNA < 500 & seurat_obj@meta.data$QC != 'Pass' & seurat_obj@meta.data$QC != 'High_MT',paste('High_MT',seurat_obj@meta.data$QC,sep = ','),seurat_obj@meta.data$QC)
table(seurat_obj[['QC']])



#Apply sctransform normalization https://satijalab.org/seurat/articles/sctransform_vignette ;single command replaces NormalizeData(), ScaleData(), and FindVariableFeatures().

seurat_obj_filt <- SCTransform(seurat_obj_filt, vars.to.regress = "percent.mt", verbose = FALSE)


# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(seurat_obj_filt), 10)

# plot variable features with and without labels
png(file = file.path(glue("{out_path}/Plots/Filtereddata_Top10_variablefeatres.png")), units = "in", width = 12, height = 12, res = 800) 
plot1 <- VariableFeaturePlot(seurat_obj_filt)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot1 + plot2
dev.off()

#linear dimensional reduction
seurat_obj_filt <- RunPCA(seurat_obj_filt, verbose = FALSE)

print(seurat_obj_filt[["pca"]], dims = 1:5, nfeatures = 5)

png(file = file.path(glue("{out_path}/Plots/Filtereddata_Genes_in_PC1_PC2.png")), units = "in", width = 12, height = 12, res = 800) 
VizDimLoadings(seurat_obj_filt, dims = 1:2, reduction = "pca")
dev.off()

png(file = file.path(glue("{out_path}/Plots/Filtereddata_PCAplot_PC1_PC2.png")), units = "in", width = 12, height = 12, res = 800) 
DimPlot(seurat_obj_filt, reduction = "pca") + NoLegend()
dev.off()

png(file = file.path(glue("{out_path}/Plots/Filtereddata_DimHeatmap_PC1_PC2.png")), units = "in", width = 12, height = 12, res = 800) 
DimHeatmap(seurat_obj_filt, dims = 1, cells = 500, balanced = TRUE)
dev.off()

png(file = file.path(glue("{out_path}/Plots/Filtereddata_DimHeatmap_PC1_PC15.png")), units = "in", width = 12, height = 12, res = 800) 
DimHeatmap(seurat_obj_filt, dims = 1:15, cells = 500, balanced = TRUE)
dev.off()

png(file = file.path(glue("{out_path}/Plots/Filtereddata_ElbowPlot.png")), units = "in", width = 12, height = 12, res = 800) 
ElbowPlot(seurat_obj_filt,ndims = 50)
dev.off()


#non-linear dimensional reduction (UMAP/tSNE)
seurat_obj_filt <- RunUMAP(seurat_obj_filt, dims = 1:30, verbose = FALSE)
seurat_obj_filt <- FindNeighbors(seurat_obj_filt, dims = 1:30, verbose = FALSE)
seurat_obj_filt <- FindClusters(seurat_obj_filt, verbose = FALSE)
png(file = file.path(glue("{out_path}/Plots/Filtereddata_UMAP.png")), units = "in", width = 12, height = 12, res = 800) 
DimPlot(seurat_obj_filt, label = TRUE,reduction = "umap")
dev.off()


##SNE
seurat_obj_filt <-RunTSNE(seurat_obj_filt, dims = 1:30, verbose = FALSE)
png(file = file.path(glue("{out_path}/Plots/Filtereddata_TNSE.png")), units = "in", width = 12, height = 12, res = 800) 
DimPlot(seurat_obj_filt, label = TRUE,reduction = "tsne")
dev.off()

#Finding differentially expressed features (cluster biomarkers)

# find all markers of cluster 2
cluster2.markers <- FindMarkers(seurat_obj_filt, ident.1 = 2)
head(cluster2.markers, n = 5)

cluster5.markers <- FindMarkers(seurat_obj_filt, ident.1 = 0, ident.2 = c(1, 19))
head(cluster5.markers, n = 5)



###-----------
monaco.ref <- celldex::MonacoImmuneData()
sce <- as.SingleCellExperiment(DietSeurat(seurat_obj_filt))
sce

monaco.main <- SingleR(test = sce,assay.type.test = 1,ref = monaco.ref,labels = monaco.ref$label.main)
monaco.fine <- SingleR(test = sce,assay.type.test = 1,ref = monaco.ref,labels = monaco.ref$label.fine)

table(monaco.main$pruned.labels)

table(monaco.fine$pruned.labels)

seurat_obj_filt@meta.data$monaco.main <- monaco.main$pruned.labels
seurat_obj_filt@meta.data$monaco.fine <- monaco.fine$pruned.labels

seurat_obj_filt <- SetIdent(seurat_obj_filt, value = "monaco.fine")

png(file = file.path(glue("{out_path}/Plots/Filtereddata_TNSE_with_genemarkers.png")), units = "in", width = 12, height = 12, res = 800) 
DimPlot(seurat_obj_filt, label = T , repel = T, label.size = 3,reduction = "tsne") + NoLegend()
dev.off()

png(file = file.path(glue("{out_path}/Plots/Filtereddata_UMAP_with_genemarkers.png")), units = "in", width = 12, height = 12, res = 800) 
DimPlot(seurat_obj_filt, label = T , repel = T, label.size = 3,reduction = "umap") + NoLegend()
dev.off()

seurat_obj_filt <- FindNeighbors(seurat_obj_filt, dims = 1:10)
FindClusters(seurat_obj_filt)
head(Idents(seurat_obj_filt), 5)




png(file = file.path(glue("{out_path}/Plots/Filtereddata_DimHeatmap_tsne_PC1_PC2.png")), units = "in", width = 12, height = 12, res = 800) 
DimHeatmap(seurat_obj_filt, dims = 1, cells = 500, balanced = TRUE,reduction = 'umap')
dev.off()


seurat_obj_filt <- RunTSNE(seurat_obj_filt, dims = 1:50, verbose = FALSE)

png(file = file.path(glue("{out_path}/Plots/Filtereddata_Dimplot_50tsne_PC1_PC2.png")), units = "in", width = 12, height = 12, res = 800) 
DimHeatmap(seurat_obj_filt, reduction = "tsne")
dev.off()




# These are now standard steps in the Seurat workflow for visualization and clustering
seurat_obj_filt <- RunPCA(seurat_obj_filt, verbose = FALSE)
seurat_obj_filt <- RunUMAP(seurat_obj_filt, dims = 1:30, verbose = FALSE)

seurat_obj_filt <- FindNeighbors(seurat_obj_filt, dims = 1:30, verbose = FALSE)
seurat_obj_filt <- FindClusters(seurat_obj_filt, verbose = FALSE)
png(file = file.path(glue("{out_path}/Plots/Filtereddata_Dimplot_50tsne_PC1_PC2.png")), units = "in", width = 12, height = 12, res = 800) 
DimPlot(seurat_obj_filt, label = TRUE)
dev.off()



# SCTranform

seurat_obj_filt <- RunPCA(seurat_obj_filt, verbose = FALSE)
seurat_obj_filt <- RunUMAP(seurat_obj_filt, dims = 1:30, verbose = FALSE)

seurat_obj_filt <- FindNeighbors(seurat_obj_filt, dims = 1:30, verbose = FALSE)
seurat_obj_filt <- FindClusters(object = seurat_obj_filt,
                               resolution = c(0.4, 0.6, 0.8, 1.0, 1.4,1.8))

Idents(object = seurat_obj_filt) <- "integrated_snn_res.1.8"
seurat_integrated <- RunUMAP(seurat_obj_filt, 
                 reduction = "pca", 
                 dims = 1:40)


png(file = file.path(glue("{out_path}/Plots/Filtereddata_Dimplot_50tsne_PC1_PC2_.1.8.png")), units = "in", width = 12, height = 12, res = 800) 
# Plot the UMAP
DimPlot(seurat_integrated,
        reduction = "umap",
        label = TRUE,
        label.size = 6)
dev.off()

png(file = file.path(glue("{out_path}/Plots/Filtereddata_Dimplot_50tsne_PC1_PC2_.1.8_2.png")), units = "in", width = 12, height = 12, res = 800) 
DimPlot(seurat_obj_filt, label = T , repel = T, label.size = 3,reduction = "umap") + NoLegend()
dev.off()